File: examples/Ruby/rtdealer.rb
# Custom routing Router to Dealer.
# Ruby version, based on the C version from
# http://zguide.zeromq.org/chapter:all#toc45
#
# libzmq: 2.1.10
# ruby: 1.9.2p180 (2011-02-18 revision 30909) [i686-linux]
# ffi-rzmq: 0.9.0
#
# @author Pavel Mitin
# @email mitin.pavel@gmail.com

require 'rubygems'
require 'ffi-rzmq'

module RTDealer
  ENDPOINT = 'ipc://routing.ipc'
  WORKER_ADDRESSES = %w(A B)
  END_MESSAGE = 'END'

  class Worker
    def run
      do_run
    ensure
      @socket.close
    end

    private

    def initialize(context, address)
      @address = address
      @socket = context.socket ZMQ::DEALER
      @socket.setsockopt ZMQ::IDENTITY, address
      @socket.connect ENDPOINT
      @total = 0
      @workload = ''
    end

    def do_run
      catch(:end) do
        loop do
          receive_workload
          handle_workload
        end
      end
      print_results
    end

    def receive_workload
      @socket.recv_string @workload
    end

    def handle_workload
      if @workload == END_MESSAGE
        throw :end
      else
        @total += 1
      end
    end

    def print_results
      p "#{@address} received: #{@total}"
    end
  end

  class Client
    def run
      send_workload
      stop_workers
    ensure
      @socket.close
    end

    private

    def initialize(context)
      @socket = context.socket ZMQ::ROUTER
      @socket.bind ENDPOINT
    end

    def send_workload
      10.times do
        address = rand(3) % 3 == 0 ? WORKER_ADDRESSES.first : WORKER_ADDRESSES.last
        @socket.send_string address, ZMQ::SNDMORE
        @socket.send_string "This is the workload"
      end
    end

    def stop_workers
      WORKER_ADDRESSES.each do |address|
        @socket.send_string address, ZMQ::SNDMORE
        @socket.send_string END_MESSAGE
      end
    end
  end
end

if $0 == __FILE__
  context = ZMQ::Context.new 1
  client = RTDealer::Client.new context
  workers = RTDealer::WORKER_ADDRESSES.map do |address|
    Thread.new { RTDealer::Worker.new(context, address).run }
  end

  sleep 1
  client.run

  workers.each &:join
  context.terminate
end

File: examples/Scala/rtdealer.scala
/**
 * Custom routing Router to Dealer.
 * Scala version, based on the C version from
 * http://zguide.zeromq.org/chapter:all#toc45
 *
 * @author Giovanni Ruggiero
 * @email giovanni.ruggiero@gmail.com
 */

import org.zeromq.ZMQ

import java.util.Arrays
import java.util.Random


/**
 * Router-to-dealer custom routing demo.
 *
 * The router, in this case the main function, uses ROUTER.  The
 * dealers, in this case the two worker threads, use DEALER.
 */
object rtdealer  {
  val NOFLAGS = 0

  /**
   * Worker runnable consumes messages until it receives an END
   * message.
   */
  class Worker(name: String) extends Runnable {

    def run() {
      val context = ZMQ.context(1)
      val socket = context.socket(ZMQ.DEALER)
      socket.setIdentity(name.getBytes())
      socket.connect("tcp://localhost:5555")

      var total = 0
			var workload = ""
			do {
				 workload = new String(socket.recv(NOFLAGS))
				total += 1
			} while (!workload.equalsIgnoreCase("END"))
			printf( "Worker %s received %d messages.\n", name, total )
      socket.close
      context.term
    }
  }

  /* Random number generator to determine message distribution. */
  val rand = new Random

  def main(args : Array[String])  {
    val context = ZMQ.context(1)
    val socket = context.socket(ZMQ.ROUTER)
    socket.bind("tcp://*:5555")

    val workerA = new Thread(new Worker("A"))
    val workerB = new Thread(new Worker("B"))
    workerA.start()
    workerB.start()

    // Wait a second for the workers to connect their sockets.
    println("Workers started, sleeping 1 second for warmup.")
    Thread.sleep(1000)

    // Send 10 tasks, scattered to A twice as often as B.
		var address = Array[Byte]()
    for (i <- 1 to 10) {
      if (rand.nextInt() % 3 == 0) { // 1/3 to B.
        address = "B".getBytes()
      } else { // 2/3 to A.
        address = "A".getBytes()
			}
      socket.send(address, ZMQ.SNDMORE)
      socket.send("This is the workload.".getBytes, NOFLAGS)
    }
    socket.send("A".getBytes, ZMQ.SNDMORE)
    socket.send("END".getBytes, NOFLAGS)

    socket.send("B".getBytes, ZMQ.SNDMORE)
    socket.send("END".getBytes, NOFLAGS)

    socket.close
    context.term
  }

}
